src/Microsoft.ML.Transforms/NAReplaceTransform.cs

```diff
@@ -481,415 +406,526 @@ private object GetDefault<T>()
+                for (int i = 0; i < _parent.ColumnPairs.Length; i++)
                 {
-                    getSrc(ref src);
-                    Host.Check(src.Length == repArray.Length);
-                    FillValues(ref src, ref dst, isNA, repArray, _repIsDefault[iinfo]);
-                };
-        }
+                    if (!inputSchema.TryGetColumnIndex(_parent.ColumnPairs[i].input, out int colSrc))
+                        throw Host.ExceptSchemaMismatch(nameof(inputSchema), "input", _parent.ColumnPairs[i].input);
+                    _parent.CheckInputColumn(inputSchema, i, colSrc);
+                    var type = inputSchema.GetColumnType(colSrc);
+                    infos[i] = new ColInfo(_parent.ColumnPairs[i].output, _parent.ColumnPairs[i].input, type);
+                }
+                return infos;
+            }
 
-        protected override bool SaveAsOnnxCore(OnnxContext ctx, int iinfo, ColInfo info, string srcVariableName, string dstVariableName)
-        {
-            DataKind rawKind;
-            var type = Infos[iinfo].TypeSrc;
-            if (type.IsVector)
-                rawKind = type.AsVector.ItemType.RawKind;
-            else if (type.IsKey)
-                rawKind = type.AsKey.RawKind;
-            else
-                rawKind = type.RawKind;
-
-            if (rawKind != DataKind.R4)
-                return false;
-
-            string opType = "Imputer";
-            var node = ctx.CreateNode(opType, srcVariableName, dstVariableName, ctx.GetNodeName(opType));
-            node.AddAttribute("replaced_value_float", Single.NaN);
-
-            if (!Infos[iinfo].TypeSrc.IsVector)
-                node.AddAttribute("imputed_value_floats", Enumerable.Repeat((float)_repValues[iinfo], 1));
-            else
+            public override RowMapperColumnInfo[] GetOutputColumns()
             {
-                if (_repIsDefault[iinfo] != null)
-                    node.AddAttribute("imputed_value_floats", (float[])_repValues[iinfo]);
-                else
-                    node.AddAttribute("imputed_value_floats", Enumerable.Repeat((float)_repValues[iinfo], 1));
+                var result = new RowMapperColumnInfo[_parent.ColumnPairs.Length];
+                for (int i = 0; i < _parent.ColumnPairs.Length; i++)
+                {
+                    InputSchema.TryGetColumnIndex(_parent.ColumnPairs[i].input, out int colIndex);
+                    Host.Assert(colIndex >= 0);
+                    var colMetaInfo = new ColumnMetadataInfo(_parent.ColumnPairs[i].output);
+                    foreach (var type in InputSchema.GetMetadataTypes(colIndex).Where(x => x.Key == MetadataUtils.Kinds.SlotNames || x.Key == MetadataUtils.Kinds.IsNormalized))
+                        Utils.MarshalInvoke(AddMetaGetter<int>, type.Value.RawType, colMetaInfo, InputSchema, type.Key, type.Value, colIndex);
+                    result[i] = new RowMapperColumnInfo(_parent.ColumnPairs[i].output, _parent._types[i], colMetaInfo);
```

Comment:
>colMetaInfo [](start = 106, length = 11)

can we just do `RowColumnUtils.GetMetadataAsRow(InputSchema, colIndex, x=> x == SlotNames || x == IsNormalized)` ?

Reply:
This is the whole purpose of this 'metadata as IRow' exercise, to avoid creating such 'identity getters' for metadata

---
In reply to: [217836483](https://github.com/dotnet/machinelearning/pull/917#discussion_r217836483) [](ancestors = 217836483)